### ISSUE TITLE
LSTM memory leak?

### ISSUE THREAD

**Issue by kennysong** [pytorch/pytorch#1536]
The following script consumes about 1 GB of memory in 100 iterations, and continually increases memory usage. It's a stripped down version of a larger file to isolate the potential leak.

Commenting out lines 33-34 stabilizes memory usage to 30 MB.

```python
import numpy as np
import torch

from torch.autograd import Variable

class PolicyNet(torch.nn.Module):
    def __init__(self, layers):
        super(PolicyNet, self).__init__()
        self.lstm = torch.nn.LSTMCell(layers[0], layers[1])
        self.linear = torch.nn.Linear(layers[1], layers[2])
        self.layers = layers

    def forward(self, x, h0, c0):
        h1, c1 = self.lstm(x, (h0, c0))
        o1 = self.linear(h1)
        return o1, h1, c1

def leak(policy_net):
    # Prepare for one forward pass, with the batch containing the entire episode
    h_n_batch = Variable(torch.zeros(1000, 32))
    c_n_batch = Variable(torch.zeros(1000, 32))
    softmax = torch.nn.Softmax()

    # Batch input to LSTM has size [num_agents, episode_len, lstm_input_size]
    input_batch = Variable(torch.zeros(2, 1000, 5))

    # Do a forward pass, and fill sum_log_probs with sum(log(p)) for each time-step
    sum_log_probs = Variable(torch.zeros(1000))
    for i in range(2):
        o_n, h_n_batch, c_n_batch = policy_net(input_batch[i], h_n_batch, c_n_batch)
        dist = softmax(o_n)
        ### Comment out L33-34 to stop memory leak ###
        for j in range(1000):
            sum_log_probs[j] = sum_log_probs[j] + torch.log(dist[j, 0])

if __name__ == '__main__':
    policy_net = PolicyNet([5, 32, 3])

    for i in range(10000):
        leak(policy_net)
        print(i)
```

This is for version `0.1.12_2` on macOS 10.12.4.

**Comment by apaszke** [comment]
Hmmm maybe it's fixed on master? I just ran it and the memory usage stays constant both on CPU and GPU.

**Comment by kennysong** [comment]
Ah interesting, does seem to be fixed on master. Do you know when the changes may make it into the next release?

**Comment by soumith** [comment]
next release is a major release. i estimate it to be end of may or early june.

**Comment by kennysong** [comment]
Hmm, alright. Will have to work with the master version for now then. 

Out of curiosity, any idea what the fix was for this memory leak?

**Comment by apaszke** [comment]
It was probably the autograd refactor that removed Variables from the graph (they were replaced with AccumulateGrad nodes). The leak was likely a reference cycle

**Comment by rasoolfa** [comment]
Hi @apaszke ,

I guess this problem still exists even on the master.
I followed the https://discuss.pytorch.org/t/tracking-down-a-suspected-memory-leak/1130/42 as well but I don't think so it is completely solved. My Pytorch version is '0.2.0+4a4d884' (got the latest version yesterday from)

Take the following as an example:

```
import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np
import torch.autograd as autograd

def Variable(data, *args, **kwargs):
    if torch.cuda.is_available():
        return autograd.Variable(data.cuda(), *args, **kwargs)
    else:
        return autograd.Variable(data, *args, **kwargs)

class LSTM_MEM_LEAK(nn.Module):
    
    def __init__(self):
 
        super(LSTM_MEM_LEAK, self).__init__()
        self.l1 = nn.Linear(256, 300)
        self.lstm =nn.LSTM(300, 400, batch_first = True, num_layers = 1)
        self.l2 = nn.Linear(400, 300)
            
    def forward(self, input):
        
        hidden = (Variable(torch.zeros(1, 16, 400)),
                Variable(torch.zeros(1, 16, 400))) 
        l1 = F.relu(self.l1(input.view(-1, 256)))
        lstm_out, h = self.lstm(l1.view(16, -1, 300), hidden)
        l2  = F.relu(self.l2(lstm_out.contiguous().view(-1, 400)))
        
        return l2    
    
net = LSTM_MEM_LEAK()
net.cuda()
input = Variable(torch.rand(16, 6000, 256))
print(input.requires_grad)
out = net(input)
```
The memory usage before out = net(input) is ~240MB but when this line runs, it goes up to 1708MB. 

Since no gradient will be computed for the input (i.e. requires_grad is False --> so no need to save parameters at the each time steps to calculate the gradient of input), I am not sure why the memory keeps increasing. 
Any comment would be highly appreciated.
